board: arm: acsip_s76s: add AcSIP S76S support

NorthernDean opened the pull request:
The AcSIP S76S is a STM32L073+SX1276+PA SoC which
supports LoRa communications.

Signed-off-by: Dean Weiten <dmw@weiten.com>

NorthernDean committed:
dts: arm: s76s: add AcSIP S76S SiP support

The AcSIP S76S is a STM32L073+SX1276+PA SoC which
supports LoRa communications.

Signed-off-by: Dean Weiten <dmw@weiten.com>

galak commented:
Was curious about the .csv files, is that purely for the docs and making the pin mgmt easier to describe?

erwango commented:
Thanks for this contribution!
I'm only commenting the form now but SiP porting as a board is unusual. On other hand I don't think we have a policy in place for SiP/modules definition.
My first idea would tend to assimilate SiP to a SoC and have a "real board" ported here, such as [this one](https://ronoth.com/products/lodev-s76s-lora-soc-development-board?variant=31608819417220). 

Then, if idea is to enable modules/SiP definitions, I'd rather put them in a new dedicated `dts/modules` section or `boards/modules`. But to me a zephyr target should remain a real board. @nashif, @MaureenHelm, @galak any opinion ?

NorthernDean commented:
> Was curious about the .csv files, is that purely for the docs and making the pin mgmt easier to describe?

Hi!

I tried to reply to the comment E-mail, but I'm not sure that it worked, so I will reprise my answer here... and I've changed my tack, (spoiler alert) I will get rid of the CSV files.  It might take me a bit, they are ugly...

Yes, I created spreadsheets for my own use, and figured they might help someone else, so I CSV'd them and put them in the doc.  I guess it's a bit atypical. 

I was trying to avoid the extra work of reformatting them into RST tables, taking the easy way out by just creating CSVs and side-loading them into the CSV.

The CSVs might be ugly (weird export from LibreOffice), but when imported into a spreadsheet, they make up what I think is a nice table that can be used to plan the embedded design. 

For this chip, many of the underlying processor (STM32L073RZ) pins are not brought out to the chip pins, and I was finding it difficult to figure out what 'L073 resources I had access to, and which ones I did not.  So, I created the spreadsheet with all the main and alternate pin assignments, then split it into three parts, so I could do a "find" and see which section things were in, instead of constantly hunting for, and often missing, things on the overall chip's datasheet.

This is actually my first tentative foray into open source contribution, so I'm not quite sure what I'm doing.  I mean, I've been doing embedded and real-time software for almost 40 years, but never contributed to this kind of project.

Well, after struggling with the "documentation github workflow", it doesn't like my CSVs.  I will fold them into the index.rst and resubmit.  It might take a bit of work, but I will be back :-)

Dean.

NorthernDean commented:
> Thanks for this contribution!
> I'm only commenting the form now but SiP porting as a board is unusual. On other hand I don't think we have a policy in place for SiP/modules definition.
> My first idea would tend to assimilate SiP to a SoC and have a "real board" ported here, such as [this one](https://ronoth.com/products/lodev-s76s-lora-soc-development-board?variant=31608819417220).
> 
> Then, if idea is to enable modules/SiP definitions, I'd rather put them in a new dedicated `dts/modules` section or `boards/modules`. But to me a zephyr target should remain a real board. @nashif, @MaureenHelm, @galak any opinion ?

Hi, 

Again, per previous comment, I tried to submit my response in an E-mail, but I'm not sure that it worked, so I'll try here.

This is my first contribution to an open source project in my 40 year career...  and I'm a bit bewildered, trying to be careful.

Anyway, I struggled with that same question.  The challenge is, supporting it as an SoC seems to be a much more complex endeavour.  If I could just move the files from one tree to the other, I'd be happy to do so.

Yes, the Ronoth LoDev board that you cite is the development system, and I'd be happy to put it in the "boards" section if there's a way to get this part into the "soc" section without killing myself.  Especially, as I say, this is my first effort in open source :-)

Dean.

NorthernDean commented:
Hi there,

Well, my branch has passed the automated checks.  

I found a Python script to convert my CSV to an RST table, but it choked on my table because it had embedded newlines in the cells (argh, that's why *I* didn't want to manually convert them)... so I hacked it and made it work.  So no more CSVs.

The question of whether this is appropriate to merge as a "board" instead of something else is _a propos_.  I seek some guidance on this issue.  Is there an easy way to drop this structure into another, more appropriate place?  It didn't seem so, from my brief look... at least not in the "soc" area, where I figured it might go.  

The point here is, that the S76S contains an STMicro STM32L073RZ processor, a Semtech SX1276 LoRa transceiver, and a power amplifier.  I didn't see an appropriate, "easy", way, to indicate this integration.

Please advise if there's a better way to do it :-)

Dean.

galak commented:
> Hi there,
> 
> Well, my branch has passed the automated checks.
> 
> I found a Python script to convert my CSV to an RST table, but it choked on my table because it had embedded newlines in the cells (argh, that's why _I_ didn't want to manually convert them)... so I hacked it and made it work. So no more CSVs.
> 
> The question of whether this is appropriate to merge as a "board" instead of something else is _a propos_. I seek some guidance on this issue. Is there an easy way to drop this structure into another, more appropriate place? It didn't seem so, from my brief look... at least not in the "soc" area, where I figured it might go.
> 
> The point here is, that the S76S contains an STMicro STM32L073RZ processor, a Semtech SX1276 LoRa transceiver, and a power amplifier. I didn't see an appropriate, "easy", way, to indicate this integration.
> 
> Please advise if there's a better way to do it :-)
> 
> Dean.

I'd say maybe a mix here between a full SoC and board port makes sense.  The main thing I'm thinking is that we introduce the "SIP" .dtsi in dts/arm/acsip/... than defines the common SIP bits and that gets used in the board dts.

We can see over time if it makes sense to refactor some the 'Kconfig' logic out of the board if we end up having multiple boards for the same SIP or we have a number of different SIPs that have support in Zephyr.

NorthernDean commented:
I am looking into splitting this up appropriately such that the AcSIP S76S "chip" SiP is supported by a dts/arm dtsi file, and then the Ronoth LoDev "board" in the board/arm section builds on top of that.

NorthernDean commented:
_Now_ I need some assistance, please.  I seems that the compliance check wants a code owner for selected files.  What do I do?  I do not have commit privilege on the repository, but the comment seems to say that I need it.  I'm not sure I want it yet :-)

martinjaeger commented:
> _Now_ I need some assistance, please.  I seems that the compliance check wants a code owner for selected files.  What do I do?  

You need to add yourself to the codeowners file for the new files (see compliance check message) in your commit in this PR.

NorthernDean commented:
> > _Now_ I need some assistance, please.  I seems that the compliance check wants a code owner for selected files.  What do I do?
> 
> You need to add yourself to the codeowners file for the new files (see compliance check message) in your commit in this PR.

Should I add myself to the codeowners for the other (new) files in my commit?

NorthernDean commented:
I hope that my submission is structured correctly.  Let me know if I need to break it up, restructure, move things etc.  I am using this to generate test code for the Ronoth LoDev, and it seems to work :-)

NorthernDean commented:
Hi, it has been some time without feedback.  Is there something else I should be doing?

galak commented:
> Hi, it has been some time without feedback. Is there something else I should be doing?

Mostly people are busy working on bugs as we try and get the 2.5 release out this week.

NorthernDean commented:
> > Hi, it has been some time without feedback. Is there something else I should be doing?
> 
> Mostly people are busy working on bugs as we try and get the 2.5 release out this week.

Ah yes, no worries, I will chill and wait then.  Thanks!

erwango commented on CODEOWNERS:
```diff
@@ -305,6 +306,7 @@
 /drivers/wifi/winc1500/                   @kludentwo
 /drivers/virtualization/		  @tbursztyka
 /dts/arc/                                 @abrodkin @ruuddw @iriszzw
+/dts/arm/acsip/                           @NorthernDean
```
@galak Can't recall the conclusion of the dev review. Is that the expected way ?

NorthernDean commented:
Is there something else I should do to fix this pull request, or promote its adoption?  This is my first time contributing, and I'd like to know if there is something wrong with the request.

erwango commented:
> Is there something else I should do to fix this pull request, or promote its adoption? This is my first time contributing, and I'd like to know if there is something wrong with the request.

@NorthernDean You're doing good, we just need some more time to review. Part of the problem is that this is a new type of board and it requires discussion and agreement to happen, which is taking more time. Please bear with us.

NorthernDean commented:
> > Is there something else I should do to fix this pull request, or promote its adoption? This is my first time contributing, and I'd like to know if there is something wrong with the request.
> 
> @NorthernDean You're doing good, we just need some more time to review. Part of the problem is that this is a new type of board and it requires discussion and agreement to happen, which is taking more time. Thanks for bearing with us.

Oh OK, no worries, if this is something new (a bit surprised but that is so cool!), then I'll not worry about it.  As time goes by, I might want to contribute in other ways :-)

erwango reviewed (changes requested):
This looks generally fine.
Only request from my part would be to split dts and board parts in 2 distinct commits.
There are other points of attention, like references to Mbed developments or .rst under `dts/` which are unusual and I'll let these to @galak and @MaureenHelm

martinjaeger reviewed (commented):
I'm wondering if we should really add this board/SIP to the `dac`, `adc` and `mem_protect` unit tests. As the underlying SOC is the same as the Nucleo L073RZ and the "normal" STM32L073RZ is probably more common, adding it to the unit tests will mainly increase CI compile time, but not really improve the test coverage.

So I would suggest to only add the board definitions and keep the unit tests as they are. What do others think?

NorthernDean commented:
> I'm wondering if we should really add this board/SIP to the `dac`, `adc` and `mem_protect` unit tests. As the underlying SOC is the same as the Nucleo L073RZ and the "normal" STM32L073RZ is probably more common, adding it to the unit tests will mainly increase CI compile time, but not really improve the test coverage.
> 
> So I would suggest to only add the board definitions and keep the unit tests as they are. What do others think?

As long as it doesn't affect use of the definitions, then that would be a good idea.  What portion of the definition drives the unit tests?  

What do other similar SoCs do?  Surely this isn't the first :-)  I was trying to re-use existing way of doing things, did I miss something?

NorthernDean commented:
> This looks generally fine.
> Only request from my part would be to split dts and board parts in 2 distinct commits.
> There are other points of attention, like references to Mbed developments or .rst under `dts/` which are unusual and I'll let these to @galak and @MaureenHelm

Hi, I was wondering about that myself.  I will split them up and submit them separately, thanks.

erwango commented:
@NorthernDean You can have multiple commits in one PR. So you can just push your 2 commits here they will be added on top of master at PR merge.

NorthernDean committed:
board: arm: ronoth_lodev: add Ronoth LoDev board support

The Ronoth LoDev in an open source board which uses
the AcSIP S76S.

Signed-off-by: Dean Weiten <dmw@weiten.com>

NorthernDean committed:
samples: drivers: dac: add Ronoth Lodev to supported boards

The Ronoth LoDev in an open source board which uses
the AcSIP S76S, which itself contains an STM32L073.

Signed-off-by: Dean Weiten <dmw@weiten.com>

NorthernDean committed:
tests: drivers: adc: add Ronoth Lodev to supported boards

The Ronoth LoDev in an open source board which uses
the AcSIP S76S, which itself contains an STM32L073.

Signed-off-by: Dean Weiten <dmw@weiten.com>

NorthernDean committed:
tests: drivers: dac: add Ronoth Lodev to supported boards

The Ronoth LoDev in an open source board which uses
the AcSIP S76S, which itself contains an STM32L073.

Signed-off-by: Dean Weiten <dmw@weiten.com>

NorthernDean committed:
tests: kernel: mem_protect: add Ronoth Lodev to supported boards

The Ronoth LoDev in an open source board which uses
the AcSIP S76S, which itself contains an STM32L073.

Signed-off-by: Dean Weiten <dmw@weiten.com>

NorthernDean commented:
> @NorthernDean You can have multiple commits in one PR. So you can just push your 2 commits here they will be added on top of master at PR merge.

@erwango Thank you, I was not familiar with the ability to stack commits inside a single PR.  I'll do that.

galak commented on dts/arm/acsip/doc/s76s.rst:
```diff
@@ -0,0 +1,251 @@
+.. _acsip_s76s:
+
```
This is probalby better kept in the boards dir for now.

NorthernDean replied:
OK, I will move it over to the boards directory.  It just kind of seemed appropriate for the SiP / SoC info to be connected to the SiP / SoC definition.

erwango reviewed (approved):
Seems my approval was removed for some reason

galak commented on boards/arm/ronoth_lodev/doc/index.rst:
```diff
@@ -0,0 +1,183 @@
+
+`Board design files`_ are available on GitHub.
+
+Some non-Zephyr development has been `posted on MBed by Steve Osborn`_.
```
This is odd to have in docs, I'd ask to remove it since its not relevant to the board from a hw point of view and not to the zephyr software.

NorthernDean replied:
Well it's not MBED code, but it's code in the MBED repository, giving more information than the (unfortunately sketchy) manufacturer's datasheet, which honestly is very thin.  But hey, I'm not hung up on it, I will take the reference out.

galak reviewed (changes requested):
I'd say move  dts/arm/acsip/doc/s76s.rst under boards/arm/ronoth_lodev/doc for now.  Also as @erwango pointed out the mbed references dont make sense in the board docs.

NorthernDean commented:
OK I hope I've done things correctly.  You will of course let me know :-)

galak commented:
@NorthernDean I rebased this to fix the merge conflicts.

ioannisg commented on boards/arm/ronoth_lodev/ronoth_lodev_defconfig:
```diff
@@ -0,0 +1,33 @@
+CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=16000000
+
+# Enable MPU
+CONFIG_ARM_MPU=y
```
@NorthernDean does the SoC implement the ARM MPU?

ioannisg replied:
Nevermind, i saw that it does; it is not defined in the SOC_SERIES, but on the SoC variant level. I would have expected it be added in the series (unless certain series variants do not implement it)

ioannisg commented on tests/kernel/mem_protect/syscalls/src/main.c:
```diff
@@ -15,7 +15,8 @@
 
 #if defined(CONFIG_BOARD_NUCLEO_F429ZI) || defined(CONFIG_BOARD_NUCLEO_F207ZG) \
-	|| defined(CONFIG_BOARD_NUCLEO_L073RZ)
+	|| defined(CONFIG_BOARD_NUCLEO_L073RZ) \
```
I wonder if we should have the SoC, here, not the Board.

NorthernDean replied:
It would make sense to have the SoC here for sure, but it appears that convention and existing code has boards here.  I had not included this at first, but the build fails without it.
